crates/biome_js_analyze/src/lint/nursery/no_vue_data_object_declaration.rs

```diff
@@ -0,0 +1,156 @@
+use biome_analyze::{Ast, FixKind, Rule, RuleDiagnostic, context::RuleContext, declare_lint_rule};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_syntax::{AnyJsExpression, AnyJsStatement, JsPropertyObjectMember, T};
+use biome_rowan::{AstNode, BatchMutationExt};
+
+use crate::JsRuleAction;
+
+declare_lint_rule! {
+    /// Enforce component's data property to be a function
+    ///
+    /// Put context and details about the rule.
+    /// As a starting point, you can take the description of the corresponding _ESLint_ rule (if any).
+    ///
+    /// Try to stay consistent with the descriptions of implemented rules.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// <script>
+    /// export default {
+    ///   /* ✗ BAD */
+    ///   data: {
+    ///     foo: null
+    ///   }
+    /// }
+    /// </script>
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js,expect_diagnostic
+    /// <script>
+    /// export default {
+    ///   /* ✓ GOOD */
+    ///   data() {
+    ///     return {
+    ///       foo: null
+    ///     }
+    ///   }
+    /// }
+    /// </script>
+    /// ```
+    ///
+    pub NoVueDataObjectDeclaration {
+        version: "next",
+        name: "noVueDataObjectDeclaration",
+        language: "js",
+        recommended: true,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+impl Rule for NoVueDataObjectDeclaration {
+    type Query = Ast<JsPropertyObjectMember>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let member = ctx.query();
+
+        if let Ok(member_name) = member.name() {
+            if let Some(name) = member_name.name() {
+                if name.trim() == "data" {
+                    if let Ok(value) = member.value() {
+                        match value {
+                            AnyJsExpression::JsIdentifierExpression(_)
+                            | AnyJsExpression::JsArrowFunctionExpression(_)
+                            | AnyJsExpression::JsFunctionExpression(_) => return None,
+                            _ => return Some(()),
+                        }
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Object declaration on 'data' property is deprecated. Using function declaration instead."
+                },
+            )
+            .note(markup! {
+                "When using the data property on a component (i.e. anywhere except on new Vue), the value must be a function that returns an object."
```

Comment:
nit: syntax clarity. I'm not sure if its worth mentioning the `new Vue()` exception at all.
```suggestion
                "When using the data property on a component (i.e. anywhere except on `new Vue()`), the value must be a function that returns an object."
```